doc/man/doveadm-log.1.in
.\" Copyright (c) 2010-2018 Dovecot authors, see the included COPYING file
.TH DOVEADM\-LOG 1 "2013-11-24" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-log \- Locate, test or reopen Dovecot\(aqs log files
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv "] " "log errors" " [" \-s
.IR min_timestamp ]
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] " "log find"
.RI [ directory ]
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] " "log reopen"
.\"-------------------------------------
.br
.BR doveadm " [" \-Dv "] " "log test"
.\"------------------------------------------------------------------------
.SH DESCRIPTION
The
.B doveadm log
.I commands
are used to locate and reopen the log files of
.BR dovecot (1).
It\(aqs also possible to test the configured targets of the
.I *log_path
settings.
.\"------------------------------------------------------------------------
@INCLUDE:global-options@
.\"------------------------------------------------------------------------
.SH COMMANDS
.SS log errors
.BR "doveadm log errors" " [" \-s
.IR min_timestamp ]
.PP
The
.B log errors
command is used to show the last  \- up to 1,000 \- errors and warnings.
If no output is generated, no errors have occurred since the last start.
.PP
.TP
.BI \-s \ min_timestamp
An integer value, representing seconds since the epoch \- also known as
Unix timestamp.
When a min_timestamp was given,
.BR doveadm (1)
will only show errors occurred since that point in time.
.\"-------------------------------------
.SS log find
.B doveadm log find
.RI [ directory ]
.PP
The
.B log find
command is used to show the location of the log files, to which
.BR dovecot (1)
sends its log messages.
If
.BR dovecot (1)
logs its messages through
.BR syslogd (8)
and
.BR doveadm (1)
could not find any log files, you can specify the
.I directory
where your syslogd writes its log files.
.\"-------------------------------------
.SS log reopen
.B doveadm log reopen
.PP
This command causes
.B doveadm
to send a SIGUSR1 signal to master process, which causes it to
reopen all log files configured in the
.IR log_path ,
.I info_log_path
and
.I debug_log_path
settings.
These settings are configured in
.IR @pkgsysconfdir@/conf.d/10\-logging.conf .
.br
The master process also signals the log process to do the same.
This is mainly useful after manually rotating the log files.
.\"-------------------------------------
.SS log test
.B doveadm log test
.PP
This command causes
.B doveadm
to write the message \(dqThis is Dovecot\(aqs
.I priority
log
.RI ( timestamp )\(dq
to the configured log files.
The used priorities are:
.BR debug ,
.BR info ,
.BR warning ,
.BR error
and
.BR fatal .
.\"------------------------------------------------------------------------
.SH EXAMPLE
This example shows how to locate the log files used by
.BR dovecot (1).
.sp
.nf
.B doveadm log find
Looking for log files from /var/log
Debug: /var/log/dovecot.debug
Info: /var/log/mail.log
Warning: /var/log/mail.log
Error: /var/log/mail.log
Fatal: /var/log/mail.log
.fi
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1)

NEWS
v2.3.13 2021-01-04	Aki Tuomi <aki.tuomi@open-xchange.com>

	* CVE-2020-24386: Specially crafted command can cause IMAP hibernate to
	  allow logged in user to access other people's emails and filesystem
	  information.
	* Metric filter and global event filter variable syntax changed to a
	  SQL-like format. See https://doc.dovecot.org/configuration_manual/event_filter/
	* auth: Added new aliases for %{variables}. Usage of the old ones is
	  possible, but discouraged.
	* auth: Removed RPA auth mechanism, SKEY auth mechanism, NTLM auth
	  mechanism and related password schemes.
	* auth: Removed passdb-sia, passdb-vpopmail and userdb-vpopmail.
	* auth: Removed postfix postmap socket
	+ auth: Added new fields for auth server events. These fields are now
	  also available for all auth events. See
	  https://doc.dovecot.org/admin_manual/list_of_events/#authentication-server
	  for details.
	+ imap-hibernate: Added imap_client_hibernated, imap_client_unhibernated
	  and imap_client_unhibernate_retried events. See
	  https://doc.dovecot.org/admin_manual/list_of_events/ for details.
	+ lib-index: Added new mail_index_recreated event. See
	  https://doc.dovecot.org/admin_manual/list_of_events/#mail-index-recreated
	+ lib-sql: Support TLS options for cassandra driver. This requires
	  cpp-driver v2.15 (or later) to work reliably.
	+ lib-storage: Missing $HasAttachment / $HasNoAttachment flags are now
	  added to existing mails if mail_attachment_detection_option=add-flags
	  and it can be done inexpensively.
	+ login proxy: Added login_proxy_max_reconnects setting (default 3) to
	  control how many reconnections are attempted.
	+ login proxy: imap/pop3/submission/managesieve proxying now supports
	  reconnection retrying on more than just connect() failure. Any error
	  except a non-temporary authentication failure will result in reconnect
	  attempts.
	- auth: Lua passdb/userdb leaks stack elements per call, eventually
	  causing the stack to become too deep and crashing the auth or
	  auth-worker process.
	- auth: SASL authentication PLAIN mechanism could be used to trigger
	  read buffer overflow. However, this doesn't seem to be exploitable in
	  any way.
	- auth: v2.3.11 regression: GSSAPI authentication fails because dovecot
	  disallows NUL bytes for it.
	- dict: Process used too much CPU when iterating keys, because each key
	  used a separate write() syscall.
	- doveadm-server: Crash could occur if logging was done outside command
	  handling. For example http-client could have done debug logging
	  afterwards, resulting in either segfault or
	  Panic: file http-client.c: line 642 (http_client_context_close):
	  assertion failed: (cctx->clients_list == NULL).
	- doveadm-server: v2.3.11 regression: Trying to connect to doveadm server
	  process via starttls assert-crashed if there were no ssl=yes listeners:
	  Panic: file master-service-ssl.c: line 22 (master_service_ssl_init):
	  assertion failed: (service->ssl_ctx_initialized).
	- fts-solr: HTTP requests may have assert-crashed:
	  Panic: file http-client-request.c: line 1232 (http_client_request_send_more):
	  assertion failed: (req->payload_input != NULL)
	- imap: IMAP NOTIFY could crash with a segmentation fault due to a bad
	  configuration that causes errors. Sending the error responses to the
	  client can cause the segmentation fault. This can for example happen
	  when several namespaces use the same mail storage location.
	- imap: IMAP NOTIFY used on a shared namespace that doesn't actually
	  exist (e.g. public namespace for a nonexistent user) can crash with a panic:
	  Panic: Leaked view for index /tmp/home/asdf/mdbox/dovecot.list.index: Opened in (null):0
	- imap: IMAP session can crash with QRESYNC extension if many changes
	  are done before asking for expunged mails since last sync.
	- imap: Process might hang indefinitely if client disconnects after
	  sending some long-running commands pipelined, for example FETCH+LOGOUT.
	- lib-compress: Mitigate crashes when configuring a not compiled in
	  compression. Errors with compression configuration now distinguish
	  between not supported and unknown.
	- lib-compression: Using xz/lzma compression in v2.3.11 could have
	  written truncated output in some situations. This would result in
	  "Broken pipe" read errors when trying to read it back.
	- lib-compression: zstd compression could have crashed in some situations:
	  Panic: file ostream.c: line 287 (o_stream_sendv_int): assertion failed: (!stream->blocking)
	- lib-dict: dict client could have crashed in some rare situations when
	  iterating keys.
	- lib-http: Fix several assert-crashes in HTTP client.
	- lib-index: v2.3.11 regression: When mails were expunged at the same
	  time as lots of new content was being saved to the cache (e.g. cache
	  file was lost and is being re-filled) a deadlock could occur with
	  dovecot.index.cache / dovecot.index.log.
	- lib-index: v2.3.11 regression: dovecot.index.cache file was being
	  purged (rewritten) too often when it had a field that hadn't been
	  accessed for over 1 month, but less than 2 months. Every cache file
	  change caused a purging in this situation.
	- lib-mail: MIME parts were not returned correctly by Dovecot MIME parser.
	  Regression caused by fixing CVE-2020-12100.
	- lib-mail: When max nested MIME parts were reached, IMAP BODYSTRUCTURE
	  was written in a way that may have caused confusion for both IMAP
	  clients and Dovecot itself when parsing it. The truncated part is now
	  written out using application/octet-stream MIME type.
	- lib-mail: v2.3.11 regression: Mail delivery / parsing crashed when the
	  10000th MIME part was message/rfc822 (or if parent was multipart/digest):
	  Panic: file message-parser.c: line 167 (message_part_append):
	  assertion failed: (ctx->total_parts_count <= ctx->max_total_mime_parts).
	- lib-oauth2: Dovecot incorrectly required oauth2 server introspection
	  reply to contain username with invalid token.

src/doveadm/doveadm-log.c
/* Copyright (c) 2010-2018 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "ioloop.h"
#include "istream.h"
#include "hash.h"
#include "str.h"
#include "strescape.h"
#include "time-util.h"
#include "master-service-private.h"
#include "master-service-settings.h"
#include "doveadm.h"
#include "doveadm-print.h"

#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <dirent.h>
#include <signal.h>
#include <sys/stat.h>

#define LAST_LOG_TYPE LOG_TYPE_PANIC
#define TEST_LOG_MSG_PREFIX "This is Dovecot's "
#define LOG_ERRORS_FNAME "log-errors"
#define LOG_TIMESTAMP_FORMAT "%b %d %H:%M:%S"

extern struct doveadm_cmd doveadm_cmd_log[];

static void ATTR_NULL(2)
cmd_log_test(int argc ATTR_UNUSED, char *argv[] ATTR_UNUSED)
{
	struct failure_context ctx;
	unsigned int i;

	master_service->log_initialized = FALSE;
	master_service->flags |= MASTER_SERVICE_FLAG_DONT_LOG_TO_STDERR;
	master_service_init_log(master_service);

	i_zero(&ctx);
	for (i = 0; i < LAST_LOG_TYPE; i++) {
		const char *prefix = failure_log_type_prefixes[i];

		/* add timestamp so that syslog won't just write
		   "repeated message" text */
		ctx.type = i;
		i_log_type(&ctx, TEST_LOG_MSG_PREFIX"%s log (%u)",
			   t_str_lcase(t_strcut(prefix, ':')),
			   (unsigned int)ioloop_time);
	}
}

static void cmd_log_reopen(int argc ATTR_UNUSED, char *argv[] ATTR_UNUSED)
{
	doveadm_master_send_signal(SIGUSR1);
}

struct log_find_file {
	const char *path;
	uoff_t size;

	/* 1 << enum log_type */
	unsigned int mask;
};

struct log_find_context {
	pool_t pool;
	HASH_TABLE(char *, struct log_find_file *) files;
};

static void cmd_log_find_add(struct log_find_context *ctx,
			     const char *path, enum log_type type)
{
	struct log_find_file *file;
	char *key;

	file = hash_table_lookup(ctx->files, path);
	if (file == NULL) {
		file = p_new(ctx->pool, struct log_find_file, 1);
		file->path = key = p_strdup(ctx->pool, path);
		hash_table_insert(ctx->files, key, file);
	}

	file->mask |= 1 << type;
}

static void
cmd_log_find_syslog_files(struct log_find_context *ctx, const char *path)
{
	struct log_find_file *file;
	DIR *dir;
	struct dirent *d;
	struct stat st;
	char *key;
	string_t *full_path;
	size_t dir_len;

	dir = opendir(path);
	if (dir == NULL) {
		i_error("opendir(%s) failed: %m", path);
		return;
	}

	full_path = t_str_new(256);
	str_append(full_path, path);
	str_append_c(full_path, '/');
	dir_len = str_len(full_path);

	while ((d = readdir(dir)) != NULL) {
		if (d->d_name[0] == '.')
			continue;

		str_truncate(full_path, dir_len);
		str_append(full_path, d->d_name);
		if (stat(str_c(full_path), &st) < 0)
			continue;

		if (S_ISDIR(st.st_mode)) {
			/* recursively go through all subdirectories */
			cmd_log_find_syslog_files(ctx, str_c(full_path));
		} else if (hash_table_lookup(ctx->files,
					     str_c(full_path)) == NULL) {
			file = p_new(ctx->pool, struct log_find_file, 1);
			file->size = st.st_size;
			file->path = key =
				p_strdup(ctx->pool, str_c(full_path));
			hash_table_insert(ctx->files, key, file);
		}
	}

	(void)closedir(dir);
}

static bool log_type_find(const char *str, enum log_type *type_r)
{
	unsigned int i;
	size_t len = strlen(str);

	for (i = 0; i < LAST_LOG_TYPE; i++) {
		if (strncasecmp(str, failure_log_type_prefixes[i], len) == 0 &&
		    failure_log_type_prefixes[i][len] == ':') {
			*type_r = i;
			return TRUE;
		}
	}
	return FALSE;
}

static void cmd_log_find_syslog_file_messages(struct log_find_file *file)
{
	struct istream *input;
	const char *line, *p;
	enum log_type type;
	int fd;

	fd = open(file->path, O_RDONLY);
	if (fd == -1)
		return;
	
	input = i_stream_create_fd_autoclose(&fd, 1024);
	i_stream_seek(input, file->size);
	while ((line = i_stream_read_next_line(input)) != NULL) {
		p = strstr(line, TEST_LOG_MSG_PREFIX);
		if (p == NULL)
			continue;
		p += strlen(TEST_LOG_MSG_PREFIX);

		/* <type> log */
		T_BEGIN {
			if (log_type_find(t_strcut(p, ' '), &type))
				file->mask |= 1 << type;
		} T_END;
	}
	i_stream_destroy(&input);
}

static void cmd_log_find_syslog_messages(struct log_find_context *ctx)
{
	struct hash_iterate_context *iter;
	struct stat st;
	char *key;
	struct log_find_file *file;

	iter = hash_table_iterate_init(ctx->files);
	while (hash_table_iterate(iter, ctx->files, &key, &file)) {
		if (stat(file->path, &st) < 0 ||
		    (uoff_t)st.st_size <= file->size)
			continue;

		cmd_log_find_syslog_file_messages(file);
	}
	hash_table_iterate_deinit(&iter);
}

static void
cmd_log_find_syslog(struct log_find_context *ctx, int argc, char *argv[])
{
	const char *log_dir;
	struct stat st;

	if (argc > 1)
		log_dir = argv[1];
	else if (stat("/var/log", &st) == 0 && S_ISDIR(st.st_mode))
		log_dir = "/var/log";
	else if (stat("/var/adm", &st) == 0 && S_ISDIR(st.st_mode))
		log_dir = "/var/adm";
	else
		return;

	printf("Looking for log files from %s\n", log_dir);
	cmd_log_find_syslog_files(ctx, log_dir);
	cmd_log_test(0, NULL);

	/* give syslog some time to write the messages to files */
	sleep(1);
	cmd_log_find_syslog_messages(ctx);
}

static void cmd_log_find(int argc, char *argv[])
{
	const struct master_service_settings *set;
	const char *log_file_path;
	struct log_find_context ctx;
	unsigned int i;

	i_zero(&ctx);
	ctx.pool = pool_alloconly_create("log file", 1024*32);
	hash_table_create(&ctx.files, ctx.pool, 0, str_hash, strcmp);

	/* first get the paths that we know are used */
	set = master_service_settings_get(master_service);
	log_file_path = set->log_path;
	if (strcmp(log_file_path, "syslog") == 0)
		log_file_path = "";
	if (*log_file_path != '\0') {
		cmd_log_find_add(&ctx, log_file_path, LOG_TYPE_WARNING);
		cmd_log_find_add(&ctx, log_file_path, LOG_TYPE_ERROR);
		cmd_log_find_add(&ctx, log_file_path, LOG_TYPE_FATAL);
	}

	if (strcmp(set->info_log_path, "syslog") != 0) {
		if (*set->info_log_path != '\0')
			log_file_path = set->info_log_path;
		if (*log_file_path != '\0')
			cmd_log_find_add(&ctx, log_file_path, LOG_TYPE_INFO);
	}

	if (strcmp(set->debug_log_path, "syslog") != 0) {
		if (*set->debug_log_path != '\0')
			log_file_path = set->debug_log_path;
		if (*log_file_path != '\0')
			cmd_log_find_add(&ctx, log_file_path, LOG_TYPE_DEBUG);
	}

	if (*set->log_path == '\0' ||
	    strcmp(set->log_path, "syslog") == 0 ||
	    strcmp(set->info_log_path, "syslog") == 0 ||
	    strcmp(set->debug_log_path, "syslog") == 0) {
		/* at least some logs were logged via syslog */
		cmd_log_find_syslog(&ctx, argc, argv);
	}

	/* print them */
	for (i = 0; i < LAST_LOG_TYPE; i++) {
		struct hash_iterate_context *iter;
		char *key;
		struct log_find_file *file;
		bool found = FALSE;

		iter = hash_table_iterate_init(ctx.files);
		while (hash_table_iterate(iter, ctx.files, &key, &file)) {
			if ((file->mask & (1 << i)) != 0) {
				printf("%s%s\n", failure_log_type_prefixes[i],
				       file->path);
				found = TRUE;
			}
		}
		hash_table_iterate_deinit(&iter);

		if (!found)
			printf("%sNot found\n", failure_log_type_prefixes[i]);
	}
	hash_table_destroy(&ctx.files);
	pool_unref(&ctx.pool);
}

static const char *t_cmd_log_error_trim(const char *orig)
{
	size_t pos;

	/* Trim whitespace from suffix and remove ':' if it exists */
	for (pos = strlen(orig); pos > 0; pos--) {
		if (orig[pos-1] != ' ') {
			if (orig[pos-1] == ':') 
				pos--;
			break;
		}
	}
	return orig[pos] == '\0' ? orig : t_strndup(orig, pos);
}

static void cmd_log_error_write(const char *const *args, time_t min_timestamp)
{
	/* <type> <timestamp> <prefix> <text> */
	const char *type_prefix = "?";
	unsigned int type;
	time_t t;

	/* find type's prefix */
	for (type = 0; type < LOG_TYPE_COUNT; type++) {
		if (strcmp(args[0], failure_log_type_names[type]) == 0) {
			type_prefix = failure_log_type_prefixes[type];
			break;
		}
	}

	if (str_to_time(args[1], &t) < 0) {
		i_error("Invalid timestamp: %s", args[1]);
		t = 0;
	}
	if (t >= min_timestamp) {
		doveadm_print(t_strflocaltime(LOG_TIMESTAMP_FORMAT, t));
		doveadm_print(t_cmd_log_error_trim(args[2]));
		doveadm_print(t_cmd_log_error_trim(type_prefix));
		doveadm_print(args[3]);
	}
}

static void cmd_log_errors(int argc, char *argv[])
{
	struct istream *input;
	const char *path, *line, *const *args;
	time_t min_timestamp = 0;
	int c, fd;

	while ((c = getopt(argc, argv, "s:")) > 0) {
		switch (c) {
		case 's':
			if (str_to_time(optarg, &min_timestamp) < 0)
				i_fatal("Invalid timestamp: %s", optarg);
			break;
		default:
			help(&doveadm_cmd_log[3]);
		}
	}
	argv += optind - 1;
	if (argv[1] != NULL)
		help(&doveadm_cmd_log[3]);

	path = t_strconcat(doveadm_settings->base_dir,
			   "/"LOG_ERRORS_FNAME, NULL);
	fd = net_connect_unix(path);
	if (fd == -1)
		i_fatal("net_connect_unix(%s) failed: %m", path);
	net_set_nonblock(fd, FALSE);

	input = i_stream_create_fd_autoclose(&fd, SIZE_MAX);

	doveadm_print_init(DOVEADM_PRINT_TYPE_FORMATTED);
	doveadm_print_formatted_set_format("%{timestamp} %{type}: %{prefix}: %{text}\n");

	doveadm_print_header_simple("timestamp");
	doveadm_print_header_simple("prefix");
	doveadm_print_header_simple("type");
	doveadm_print_header_simple("text");

	while ((line = i_stream_read_next_line(input)) != NULL) T_BEGIN {
		args = t_strsplit_tabescaped(line);
		if (str_array_length(args) == 4)
			cmd_log_error_write(args, min_timestamp);
		else {
			i_error("Invalid input from log: %s", line);
			doveadm_exit_code = EX_PROTOCOL;
		}
	} T_END;
	i_stream_destroy(&input);
}

struct doveadm_cmd doveadm_cmd_log[] = {
	{ cmd_log_test, "log test", "" },
	{ cmd_log_reopen, "log reopen", "" },
	{ cmd_log_find, "log find", "[<dir>]" },
};

struct doveadm_cmd_ver2 doveadm_cmd_log_errors_ver2 = {
	.name = "log errors",
	.usage = "[-s <min_timestamp>]",
	.old_cmd = cmd_log_errors,
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('s', "since", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAMS_END
};

void doveadm_register_log_commands(void)
{
	unsigned int i;

	doveadm_cmd_register_ver2(&doveadm_cmd_log_errors_ver2);

	for (i = 0; i < N_ELEMENTS(doveadm_cmd_log); i++)
		doveadm_register_cmd(&doveadm_cmd_log[i]);
}
